MySQL failed to start in Gramine

<!-- Thanks for reporting a bug.

Please provide as much information as possible.
Paste any logs using three backticks (```).

IMPORTANT: please do not post general questions here;
if you have questions, please use our mailing list: users@gramineproject.io
-->

## Description of the problem
MySQL failed to start in Gramine via both direct and sgx modes. The MySQL error message looks like the following,
```
021-11-24T09:16:41.793587Z 0 [ERROR] Fatal error: Can't open and lock privilege tables: Table storage engine for 'user' doesn't have this option
2021-11-24T09:16:41.793616Z 0 [ERROR] Fatal error: Failed to initialize ACL/grant/time zones structures or failed to remove temporary table files.
2021-11-24T09:16:41.793739Z 0 [ERROR] Aborting
```
The full MySQL log will be attached in the following section.

This MySQL instance was initialized and started w/o Gramine. Everything seemed fine. Then I tried to bring up the instance in gramine-direct and gramine-sgx. Both direct and sgx modes showed the same error message. It might be worth mentioning that if I append "--skip-grant-tables" to MySQL command line, the database instance can be brought up and user tables (not system tables) seemed to work fine.

## Steps to reproduce
<!-- NOTE: please specify the exact commit ID on which you reproduced the issue

### PLEASE ENSURE THAT THE ISSUE REPRODUCES ON THE CURRENT MASTER BRANCH ###

-->
1. Download the MySQL binary tarball.

1.1 Go to the page following page
https://www.percona.com/downloads/Percona-Server-5.7/LATEST/

1.2 Select Version:
```
Percona Server for MySQL 5.7.35-38
```
1.3 Select Software:
```
Linux - Generic
```
1.4 Select Hardware:
```
x86_64
```
1.5 Download
```
Percona-Server-5.7.35-38-Linux.x86_64.glibc2.12.tar.gz
```
1.6 Extract the tarball. In my environment, it is 
/work/shenyc/mysql/workplace/Percona-Server-5.7.35-38-Linux.x86_64.glibc2.12

2. Configure, initialized and start MySQL w/o Gramine.

2.1 Create a my.cnf file (/work/shenyc/mysql/workplace/my.cnf in my case), and fill with the following content:
```
[mysqld]

user = root
basedir = /work/shenyc/mysql/workplace/Percona-Server-5.7.35-38-Linux.x86_64.glibc2.12
datadir = /work/shenyc/mysql/workplace/datadir
tmpdir = /work/shenyc/mysql/workplace/tmpdir

socket = /work/shenyc/mysql/workplace/mysql.sock
pid-file = /work/shenyc/mysql/workplace/mysql.pid
```

2.2 Create datadir and tmpdir
```
mkdir -p /work/shenyc/mysql/workplace/datadir /work/shenyc/mysql/workplace/tmpdir
```
2.3 Initialize:
```
/work/shenyc/mysql/workplace/Percona-Server-5.7.35-38-Linux.x86_64.glibc2.12/bin/mysqld --defaults-file=/work/shenyc/mysql/workplace/my.cnf --initialize-insecure
```
2.4 Start:
```
/work/shenyc/mysql/workplace/Percona-Server-5.7.35-38-Linux.x86_64.glibc2.12/bin/mysqld --defaults-file=/work/shenyc/mysql/workplace/my.cnf
```
3. Configure and try to bring up MySQL w/ Gramine.

3.1 Create a manifest. The manifest file I used is called mysql.manifest. The content is shown below.
```
fs.mount.lib.type = "chroot"
fs.mount.lib.path = "/lib"

fs.mount.lib.uri = "file:/work/shenyc/dist/gramine/lib/x86_64-linux-gnu/gramine/runtime/glibc"
sgx.allowed_files.lib = "file:/work/shenyc/dist/gramine/lib/x86_64-linux-gnu/gramine/runtime/glibc"
loader.entrypoint = "file:/work/shenyc/dist/gramine/lib/x86_64-linux-gnu/gramine/libsysdb.so"

fs.mount.usr.type = "chroot"
fs.mount.usr.path = "/usr"
fs.mount.usr.uri = "file:/usr"
sgx.allowed_files.usr = "file:/usr"

fs.mount.lib64.type = "chroot"
fs.mount.lib64.path = "/lib64"
fs.mount.lib64.uri = "file:/usr/lib/x86_64-linux-gnu"
sgx.allowed_files.lib64 = "file:/usr/lib/x86_64-linux-gnu"

loader.env.LD_LIBRARY_PATH = "/work/shenyc/mysql/workplace/Percona-Server-5.7.35-38-Linux.x86_64.glibc2.12/lib/private:/lib:/lib64:/usr/lib:/usr/lib64"

libos.entrypoint = "/work/shenyc/mysql/workplace/Percona-Server-5.7.35-38-Linux.x86_64.glibc2.12/bin/mysqld"

loader.insecure__use_cmdline_argv = true
loader.insecure__disable_aslr = true

sgx.enclave_size = "4G"
sgx.nonpie_binary = true
sys.insecure__allow_eventfd = true
sys.enable_sigterm_injection = true

fs.mount.etc.type = "chroot"
fs.mount.etc.path = "/etc"
fs.mount.etc.uri = "file:/etc"
sgx.allowed_files.etc = "file:/etc"

fs.mount.var.type = "chroot"
fs.mount.var.path = "/var"
fs.mount.var.uri = "file:/var"
sgx.allowed_files.var = "file:/var"

fs.mount.work.type = "chroot"
fs.mount.work.path = "/work"
fs.mount.work.uri = "file:/work"
sgx.allowed_files.work = "file:/work"

loader.log_level = "error"

libos.check_invalid_pointers = false
sgx.preheat_enclave = true

sgx.thread_num = 32

sgx.require_avx = true
sgx.require_avx512 = true
```

3.2 Build Gramine and install to
/work/shenyc/dist/gramine

3.3 Start MySQL
```
env PYTHONPATH=/work/shenyc/dist/gramine/lib/python3.8/site-packages /work/shenyc/dist/gramine/bin/gramine-direct mysql --defaults-file=/work/shenyc/mysql/workplace/my.cnf
```
## Expected results
The MySQL server continues running in foreground. The last lines of the messages shows "mysqld: ready for connections.".

## Actual results
The MySQL server exited leaving the following messages. The message ended up with "mysql: Shutdown complete".
```
2021-11-24T09:16:41.674275Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
2021-11-24T09:16:41.674429Z 0 [Note] --secure-file-priv is set to NULL. Operations related to importing and exporting data are disabled
2021-11-24T09:16:41.674454Z 0 [Note] mysql (mysqld 5.7.35-38) starting as process 1 ...
2021-11-24T09:16:41.678757Z 0 [Note] InnoDB: PUNCH HOLE support available
2021-11-24T09:16:41.678795Z 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
2021-11-24T09:16:41.678806Z 0 [Note] InnoDB: Uses event mutexes
2021-11-24T09:16:41.678816Z 0 [Note] InnoDB: GCC builtin __atomic_thread_fence() is used for memory barrier
2021-11-24T09:16:41.678826Z 0 [Note] InnoDB: Compressed tables use zlib 1.2.3
2021-11-24T09:16:41.678836Z 0 [Note] InnoDB: Using Linux native AIO
2021-11-24T09:16:41.679043Z 0 [Note] InnoDB: Number of pools: 1
2021-11-24T09:16:41.679144Z 0 [Note] InnoDB: Using CPU crc32 instructions
2021-11-24T09:16:41.679581Z 0 [ERROR] InnoDB: Linux Native AIO interface is not supported on this platform. Please check your OS documentation and install appropriate binary of InnoDB.
2021-11-24T09:16:41.679596Z 0 [Note] InnoDB: You can disable Linux Native AIO by setting innodb_use_native_aio = 0 in my.cnf
2021-11-24T09:16:41.679606Z 0 [Warning] InnoDB: Linux Native AIO disabled.
2021-11-24T09:16:41.680433Z 0 [Note] InnoDB: Initializing buffer pool, total size = 128M, instances = 1, chunk size = 128M
2021-11-24T09:16:41.684282Z 0 [Note] InnoDB: Completed initialization of buffer pool
2021-11-24T09:16:41.686341Z 0 [Warning] InnoDB: Setting thread 3 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.686544Z 0 [Warning] InnoDB: Setting thread 4 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.686784Z 0 [Warning] InnoDB: Setting thread 5 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.687112Z 0 [Warning] InnoDB: Setting thread 6 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.687295Z 0 [Warning] InnoDB: Setting thread 7 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.687621Z 0 [Warning] InnoDB: Setting thread 8 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.687795Z 0 [Warning] InnoDB: Setting thread 9 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.688177Z 0 [Warning] InnoDB: Setting thread 10 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.688312Z 0 [Warning] InnoDB: Setting thread 11 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.688710Z 0 [Warning] InnoDB: Setting thread 12 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.688897Z 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority().
2021-11-24T09:16:41.700677Z 0 [Note] InnoDB: Crash recovery did not find the parallel doublewrite buffer at /work/shenyc/mysql/workplace/datadir/xb_doublewrite
2021-11-24T09:16:41.701398Z 0 [Note] InnoDB: Highest supported file format is Barracuda.
2021-11-24T09:16:41.719159Z 0 [Note] InnoDB: Created parallel doublewrite buffer at /work/shenyc/mysql/workplace/datadir/xb_doublewrite, size 3932160 bytes
2021-11-24T09:16:41.728375Z 0 [Note] InnoDB: Creating shared tablespace for temporary tables
2021-11-24T09:16:41.728475Z 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...
2021-11-24T09:16:41.765859Z 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.
2021-11-24T09:16:41.766897Z 0 [Note] InnoDB: 96 redo rollback segment(s) found. 96 redo rollback segment(s) are active.
2021-11-24T09:16:41.766930Z 0 [Note] InnoDB: 32 non-redo rollback segment(s) are active.
2021-11-24T09:16:41.768160Z 0 [Warning] InnoDB: Setting thread 19 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.768782Z 0 [Warning] InnoDB: Setting thread 20 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.769196Z 0 [Warning] InnoDB: Setting thread 21 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.769585Z 0 [Warning] InnoDB: Setting thread 22 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.769744Z 0 [Note] InnoDB: Percona XtraDB (http://www.percona.com) 5.7.35-38 started; log sequence number 2758528
2021-11-24T09:16:41.770023Z 0 [Warning] InnoDB: Setting thread 23 nice to 0 failed, current nice 0, errno 13
2021-11-24T09:16:41.770415Z 0 [Note] InnoDB: Loading buffer pool(s) from /work/shenyc/mysql/workplace/datadir/ib_buffer_pool
2021-11-24T09:16:41.772178Z 0 [Note] Plugin 'FEDERATED' is disabled.
2021-11-24T09:16:41.779216Z 0 [Note] InnoDB: Buffer pool(s) load completed at 211124  9:16:41
2021-11-24T09:16:41.781507Z 0 [Note] Found ca.pem, server-cert.pem and server-key.pem in data directory. Trying to enable SSL support using them.
2021-11-24T09:16:41.781534Z 0 [Note] Skipping generation of SSL certificates as certificate files are present in data directory.
2021-11-24T09:16:41.781554Z 0 [Warning] A deprecated TLS version TLSv1 is enabled. Please use TLSv1.2 or higher.
2021-11-24T09:16:41.781570Z 0 [Warning] A deprecated TLS version TLSv1.1 is enabled. Please use TLSv1.2 or higher.
2021-11-24T09:16:41.790620Z 0 [Warning] CA certificate ca.pem is self signed.
2021-11-24T09:16:41.790672Z 0 [Note] Skipping generation of RSA key pair as key files are present in data directory.
2021-11-24T09:16:41.791187Z 0 [Note] Server hostname (bind-address): '*'; port: 3306
2021-11-24T09:16:41.791219Z 0 [Note] IPv6 is available.
2021-11-24T09:16:41.791234Z 0 [Note]   - '::' resolves to '::';
2021-11-24T09:16:41.791291Z 0 [Note] Server socket created on IP: '::'.
2021-11-24T09:16:41.792591Z 0 [Warning] Insecure configuration for --pid-file: Location '/work/shenyc' in the path is accessible to all OS users. Consider choosing a different directory.
2021-11-24T09:16:41.793587Z 0 [ERROR] Fatal error: Can't open and lock privilege tables: Table storage engine for 'user' doesn't have this option
2021-11-24T09:16:41.793616Z 0 [ERROR] Fatal error: Failed to initialize ACL/grant/time zones structures or failed to remove temporary table files.
2021-11-24T09:16:41.793739Z 0 [ERROR] Aborting

2021-11-24T09:16:41.793788Z 0 [Note] Binlog end
2021-11-24T09:16:41.793855Z 0 [Note] Shutting down plugin 'ngram'
2021-11-24T09:16:41.793874Z 0 [Note] Shutting down plugin 'ARCHIVE'
2021-11-24T09:16:41.793886Z 0 [Note] Shutting down plugin 'BLACKHOLE'
2021-11-24T09:16:41.793898Z 0 [Note] Shutting down plugin 'partition'
2021-11-24T09:16:41.793910Z 0 [Note] Shutting down plugin 'MyISAM'
2021-11-24T09:16:41.793925Z 0 [Note] Shutting down plugin 'CSV'
2021-11-24T09:16:41.793937Z 0 [Note] Shutting down plugin 'MEMORY'
2021-11-24T09:16:41.793949Z 0 [Note] Shutting down plugin 'MRG_MYISAM'
2021-11-24T09:16:41.793962Z 0 [Note] Shutting down plugin 'INNODB_TABLESPACES_SCRUBBING'
2021-11-24T09:16:41.793973Z 0 [Note] Shutting down plugin 'INNODB_TABLESPACES_ENCRYPTION'
2021-11-24T09:16:41.793985Z 0 [Note] Shutting down plugin 'INNODB_SYS_VIRTUAL'
2021-11-24T09:16:41.793998Z 0 [Note] Shutting down plugin 'INNODB_CHANGED_PAGES'
2021-11-24T09:16:41.794010Z 0 [Note] Shutting down plugin 'INNODB_SYS_DATAFILES'
2021-11-24T09:16:41.794022Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLESPACES'
2021-11-24T09:16:41.794038Z 0 [Note] Shutting down plugin 'INNODB_SYS_FOREIGN_COLS'
2021-11-24T09:16:41.794050Z 0 [Note] Shutting down plugin 'INNODB_SYS_FOREIGN'
2021-11-24T09:16:41.794062Z 0 [Note] Shutting down plugin 'INNODB_SYS_FIELDS'
2021-11-24T09:16:41.794073Z 0 [Note] Shutting down plugin 'INNODB_SYS_COLUMNS'
2021-11-24T09:16:41.794086Z 0 [Note] Shutting down plugin 'INNODB_SYS_INDEXES'
2021-11-24T09:16:41.794098Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLESTATS'
2021-11-24T09:16:41.794109Z 0 [Note] Shutting down plugin 'INNODB_SYS_TABLES'
2021-11-24T09:16:41.794121Z 0 [Note] Shutting down plugin 'INNODB_FT_INDEX_TABLE'
2021-11-24T09:16:41.794132Z 0 [Note] Shutting down plugin 'INNODB_FT_INDEX_CACHE'
2021-11-24T09:16:41.794144Z 0 [Note] Shutting down plugin 'INNODB_FT_CONFIG'
2021-11-24T09:16:41.794156Z 0 [Note] Shutting down plugin 'INNODB_FT_BEING_DELETED'
2021-11-24T09:16:41.794167Z 0 [Note] Shutting down plugin 'INNODB_FT_DELETED'
2021-11-24T09:16:41.794179Z 0 [Note] Shutting down plugin 'INNODB_FT_DEFAULT_STOPWORD'
2021-11-24T09:16:41.794190Z 0 [Note] Shutting down plugin 'INNODB_METRICS'
2021-11-24T09:16:41.794202Z 0 [Note] Shutting down plugin 'INNODB_TEMP_TABLE_INFO'
2021-11-24T09:16:41.794215Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_POOL_STATS'
2021-11-24T09:16:41.794226Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_PAGE_LRU'
2021-11-24T09:16:41.794238Z 0 [Note] Shutting down plugin 'INNODB_BUFFER_PAGE'
2021-11-24T09:16:41.794251Z 0 [Note] Shutting down plugin 'INNODB_CMP_PER_INDEX_RESET'
2021-11-24T09:16:41.794263Z 0 [Note] Shutting down plugin 'INNODB_CMP_PER_INDEX'
2021-11-24T09:16:41.794274Z 0 [Note] Shutting down plugin 'INNODB_CMPMEM_RESET'
2021-11-24T09:16:41.794289Z 0 [Note] Shutting down plugin 'INNODB_CMPMEM'
2021-11-24T09:16:41.794300Z 0 [Note] Shutting down plugin 'INNODB_CMP_RESET'
2021-11-24T09:16:41.794312Z 0 [Note] Shutting down plugin 'INNODB_CMP'
2021-11-24T09:16:41.794325Z 0 [Note] Shutting down plugin 'INNODB_LOCK_WAITS'
2021-11-24T09:16:41.794338Z 0 [Note] Shutting down plugin 'INNODB_LOCKS'
2021-11-24T09:16:41.794349Z 0 [Note] Shutting down plugin 'INNODB_TRX'
2021-11-24T09:16:41.794360Z 0 [Note] Shutting down plugin 'XTRADB_ZIP_DICT_COLS'
2021-11-24T09:16:41.794372Z 0 [Note] Shutting down plugin 'XTRADB_ZIP_DICT'
2021-11-24T09:16:41.794384Z 0 [Note] Shutting down plugin 'XTRADB_RSEG'
2021-11-24T09:16:41.794395Z 0 [Note] Shutting down plugin 'XTRADB_INTERNAL_HASH_TABLES'
2021-11-24T09:16:41.794407Z 0 [Note] Shutting down plugin 'XTRADB_READ_VIEW'
2021-11-24T09:16:41.794418Z 0 [Note] Shutting down plugin 'InnoDB'
2021-11-24T09:16:41.794750Z 0 [Note] InnoDB: FTS optimize thread exiting.
2021-11-24T09:16:41.795056Z 0 [Note] InnoDB: Starting shutdown...
2021-11-24T09:16:41.895593Z 0 [Note] InnoDB: Dumping buffer pool(s) to /work/shenyc/mysql/workplace/datadir/ib_buffer_pool
2021-11-24T09:16:41.896216Z 0 [Note] InnoDB: Buffer pool(s) dump completed at 211124  9:16:41
2021-11-24T09:16:42.896812Z 0 [Note] InnoDB: Waiting for page_cleaner to finish flushing of buffer pool
2021-11-24T09:16:43.908708Z 0 [Note] InnoDB: Shutdown completed; log sequence number 2758547
2021-11-24T09:16:43.914234Z 0 [Note] InnoDB: Removed temporary tablespace data file: "ibtmp1"
2021-11-24T09:16:43.914294Z 0 [Note] Shutting down plugin 'PERFORMANCE_SCHEMA'
2021-11-24T09:16:43.914447Z 0 [Note] Shutting down plugin 'sha256_password'
2021-11-24T09:16:43.914480Z 0 [Note] Shutting down plugin 'mysql_native_password'
2021-11-24T09:16:43.914893Z 0 [Note] Shutting down plugin 'binlog'
2021-11-24T09:16:43.915806Z 0 [Note] mysql: Shutdown complete
```
## Additional information
<!-- if applicable, uncomment and fill this section -->
1. The MySQL distribution used is Percona 5.7.35-38. Oracle version of MySQL server 5.7.35 was also tried. Similar error messages were seen.
2. I tried to turn on Gramine trace level log. The log was filled with lines like '[P1:T1:mysqld] warning: Unsupported system call times', and grew up to >50GB before I had to kill the test. So I can not attach a Gramine log here.

--- a/LibOS/shim/src/fs/chroot/fs.c
+++ b/LibOS/shim/src/fs/chroot/fs.c
@@ -343,6 +343,7 @@ static int chroot_istat(struct shim_inode* inode, struct stat* buf) {
     lock(&inode->lock);
     buf->st_mode = inode->type | inode->perm;
     buf->st_size = inode->size;
+    buf->st_dev = hash_str(inode->mount->uri);
     /*
      * Pretend `nlink` is 2 for directories (to account for "." and ".."), 1 for other files.
      *
@@ -367,11 +368,7 @@ static int chroot_stat(struct shim_dentry* dent, struct stat* buf) {
     }
 
     ret = chroot_istat(dent->inode, buf);
-    if (ret < 0)
-        goto out;
 
-    buf->st_dev = hash_str(dent->mount->uri);
-    ret = 0;
 out:
     unlock(&dent->lock);
     return ret;

--- a/LibOS/shim/src/fs/tmpfs/fs.c
+++ b/LibOS/shim/src/fs/tmpfs/fs.c
@@ -158,6 +158,8 @@ static int tmpfs_stat(struct shim_dentry* dent, struct stat* buf) {
     buf->st_ctime = data->ctime;
     buf->st_mtime = data->mtime;
     buf->st_atime = data->atime;
+    /* TODO: change to `hash_str(dent->inode->mount->uri)` once tmpfs supports inodes. */
+    buf->st_dev = 2;
     ret = 0;
 
 out:

